rc/tools/go/gopls.kak

```diff
@@ -0,0 +1,116 @@
+# gopls.kak: gopls bindings for kakoune
+
+define-command -params 1 -docstring %{
+gopls <command>: gopls command wrapper for kakoune
+
+All commands are forwarded to gopls utility
+Available commands are:
+    format
+    imports
+    definition
+    references
+} -shell-script-candidates %{
+    printf "format\nimports\ndefinition\nreferences\n"
+} \
+gopls %{
+    require-module gopls
+    evaluate-commands %sh{
+        case "$1" in
+        format|imports)
+            printf %s\\n "gopls-cmd $1"
+            ;;
+        definition)
+            printf %s\\n "gopls-def"
+            ;;
+        references)
+            printf %s\\n "gopls-ref"
+            ;;
+        *)
+            printf "fail Unknown gopls command '%s'\n" "$1"
+            exit
+            ;;
+        esac
+    }
+}
+
+provide-module gopls %§
+
+evaluate-commands %sh{
+    if ! command -v gopls > /dev/null 2>&1; then
+        echo "fail Please install gopls or add to PATH!"
+    fi
+}
+
+# Temp dir prepatation
+declare-option -hidden str gopls_tmp_dir
+define-command -hidden -params 0 gopls-prepare %{
+    evaluate-commands %sh{
+        dir=$(mktemp -d "${TMPDIR:-/tmp}"/kak-gopls.XXXXXXXX)
+        printf %s\\n "set-option buffer gopls_tmp_dir ${dir}"
+    }
+}
+
+# gopls format/imports
+define-command -hidden -params 1 gopls-cmd %{
+    gopls-prepare
+    evaluate-commands %sh{
+        dir=${kak_opt_gopls_tmp_dir}
+        eval "gopls $1 -w ${kak_buffile} 2> '${dir}/stderr'"
+        if [ $? -ne 0 ]; then
+            # show error messages in *debug* buffer
+            printf %s\\n "echo -debug %file{ '${dir}/stderr' }"
+        fi
+        rm -r "${dir}"
+    }
+    edit!
+}
+
+# gopls definition
+define-command -hidden -params 0 gopls-def %{
+    gopls-prepare
```

Comment:
The temporary directory+file seems unnecessary for this command. We can use a shell variable instead.

Reply:
Indeed, rewritten.